src/persist-client/src/batch.rs

```diff
@@ -307,12 +264,234 @@ where
+                            .elements()
+                            .iter()
+                            .map(|x| u64::from_le_bytes(T::encode(x)))
+                            .collect::<Vec<_>>(),
+                    ),
+                    Antichain::from(
+                        desc.upper()
+                            .elements()
+                            .iter()
+                            .map(|x| u64::from_le_bytes(T::encode(x)))
+                            .collect::<Vec<_>>(),
+                    ),
+                    Antichain::from(
+                        desc.since()
+                            .elements()
+                            .iter()
+                            .map(|x| u64::from_le_bytes(T::encode(x)))
+                            .collect::<Vec<_>>(),
+                    ),
+                );
+
+                let batch = BlobTraceBatchPart {
+                    desc: encoded_desc.clone(),
+                    updates: vec![updates],
+                    index,
+                };
+
+                let buf = mz_ore::task::spawn_blocking(
+                    || "batch::encode_part",
+                    move || {
+                        let mut buf = Vec::new();
+                        batch.encode(&mut buf);
+                        Bytes::from(buf)
+                    },
+                )
+                .instrument(debug_span!("batch::encode_part"))
+                .await
+                .expect("part encode task failed");
+
+                let () = retry_external("batch::set", || async {
+                    blob.set(
+                        Instant::now() + FOREVER,
+                        &blob_key,
+                        Bytes::clone(&buf),
+                        Atomicity::RequireAtomic,
+                    )
+                    .await
+                })
+                .instrument(trace_span!("batch::set", payload_len = buf.len()))
+                .await;
+            }
+            .instrument(write_span),
+        );
+        self.writing_parts.push_back((key, handle));
+
+        while self.writing_parts.len() > self.max_outstanding {
+            let (key, handle) = self
+                .writing_parts
+                .pop_front()
+                .expect("pop failed when len was just > some usize");
+            let () = handle
+                .instrument(debug_span!("batch::max_outstanding"))
+                .await
+                .expect("part upload task failed");
+            self.finished_parts.push(key);
+        }
+    }
+
+    async fn finish(self) -> Vec<String> {
+        let mut keys = self.finished_parts;
+        if !self.writing_parts.is_empty() {
+            async {
+                for (key, handle) in self.writing_parts {
+                    let () = handle.await.expect("part upload task failed");
+                    keys.push(key);
+                }
+            }
+            .instrument(debug_span!("batch::upload"))
```

Comment:
I think I get your motivation for this: we only log when we have to wait on pending uploads. But this could also be an `#[instrument]`, on the method to keep it simpler. Feel free to keep as is, though. Maybe use `batch::finish_upload` as the name?

Reply:
done and done! you're right that I was over-thinking this